File: kotlin/62-Unique-Paths.kt
class Solution {
    fun uniquePaths(m: Int, n: Int): Int {
      val dp = Array(m + 1) { IntArray(n + 1) { 0 } }
        val indexM = m - 1
        val indexN = n - 1
        for (i in indexM downTo 0) {
            for (j in indexN downTo 0) {
                if (i == indexM && j == indexN) {
                    dp[i][j] = 1
                } else {
                    dp[i][j] = dp[i + 1][j] + dp[i][j + 1]
                }
            }
        }
        return dp[0][0]
    }
}

File: rust/62-Unique-Paths.rs
impl Solution {
    pub fn unique_paths(m: i32, n: i32) -> i32 {
        let mut bottom = vec![1; n as usize];

        for _ in 0..m - 1 {
            let mut top = vec![1; n as usize];
            for c in (0..n - 1).rev().map(|c| c as usize) {
                top[c] = bottom[c] + top[c + 1];
            }

            bottom = top;
        }

        bottom[0]
    }
}
